doc/lspconfig.txt

```diff
@@ -1,242 +1,628 @@
+  `vim.fn.glob`. See `:help vim.fn.glob` for details.
+>
+   root_dir = util.root_pattern('pyproject.toml', 'requirements.txt')
+<
+- `util.find_git_ancestor`: locates the first parent directory containing a
+  `.git` directory.
+>
+   root_dir = util.find_git_ancestor
 
+- `util.find_node_modules_ancestor`:  locates the first parent directory
+  containing a `node_modules` directory.
+>
+   root_dir = util.find_node_modules_ancestor
+<
 
-Please see the [wiki](https://github.com/neovim/nvim-lspconfig/wiki) for
-additional topics, including:
-* [Installing language servers automatically](https://github.com/neovim/nvim-lspconfig/wiki/Installing-language-servers-automatically)
-* [Snippets support](https://github.com/neovim/nvim-lspconfig/wiki/Snippets)
-* [Project local settings](https://github.com/neovim/nvim-lspconfig/wiki/Project-local-settings)
-* [Recommended plugins for enhanced language server features](https://github.com/neovim/nvim-lspconfig/wiki/Language-specific-plugins)
+- `util.find_package_json_ancestor`:  locates the first parent directory
+  containing a `package.json`.
+>
+   root_dir = util.find_json_ancestor
+<
 ==============================================================================
-11. Windows                                                *lspconfig-windows*
+COMPOSITIONAL ROOT DIRECTORIES                      *lspconfig-root-composition*
 
+The `root_dir` key in `config` and `setup` can hold any function of the form
+>
+  function custom_root_dir(filename, bufnr)
+  returns nil | string
+>
+This allows for rich composition of root directory patterns which is necessary
+for some project structures. A worked example for kotlin:
+
+>
+  local root_files = {
+    'settings.gradle', -- Gradle (multi-project)
+    'settings.gradle.kts', -- Gradle (multi-project)
+    'build.xml', -- Ant
+    'pom.xml', -- Maven
+  }
+
+  local fallback_root_files = {
+    'build.gradle', -- Gradle
+    'build.gradle.kts', -- Gradle
+  }
+  root_dir = function(fname)
+    local primary = util.root_pattern(unpack(root_files))(fname)
+    local fallback = util.root_pattern(unpack(fallback_root_files))(fname)
+    return primary or fallback
+  end
+<
+Browsing the source of the default configurations is recommended.
 
-In order for neovim to launch certain executables on Windows, it must append
-`.cmd` to the command name. To work around this, manually append `.cmd` to the
-entry `cmd` in a given plugin's setup{} call.
 ==============================================================================
-12. Contributions                                    *lspconfig-contributions*
+SINGLE FILE SUPPORT                              *lspconfig-single-file-support*
+
+Language servers require each project to have a `root` in order to provide
+many features that require cross-file indexing.
+
+Some servers support not passing a root directory as a proxy for single file
+mode under which cross-file features may be degraded. While the built-in
```

Comment:
or, in fact, are capable of creating their own project structure from the current file...

Reply:
🙃